### ISSUE TITLE
Rerun logic doesn't handle biblatex

### ISSUE THREAD

**Issue by malramsay64** [tectonic-typesetting/tectonic#195]
When compiling a simple document (text.tex) using biblatex for managing citations, the file text.run.xml causes a TeX rerun to be required until it hits the limit of 6 passes. When querying the access with the `_dump_access_info` function, the read and written hashes are different. Over the course of the 6 repetitions, the read hash rqemains the same, while the write hash changes after the first TeX iteration after BibTeX has been run. 

Additionally the test.bbl file has a different read and write hash, which remains consistent throughout all the TeX runs.

There are no issues with the resulting output file, just a minor issue with the stopping criterion. I am not familiar enough rust to trace this any deeper.

```tex
% test.tex
\documentclass[12pt]{article}

\usepackage[backend=bibtex]{biblatex}
\bibliography{test}

\title{test}
\author{me}

\begin{document}
\maketitle

Test cite \cite{test}

\printbibliography
\end{document}
```

```bibtex
# test.bib
@article{test,
    author={},
    title={},
}
```

**Comment by pkgw** [comment]
Thanks for reporting! There hasn't been much work on the `biblatex` support, so I am not surprised that there are problems, but hopefully this can be improved.

**Comment by TheZoq2** [comment]
I also have a similar issue, at least I think I do.

Whenever I run tectonic, it runs `BibTeX`, then says `Rerunning TeX because bibtex was run ...`, even if the bib file hasn't changed.

If I wanted to try and fix this, where should I start looking?

**Comment by pkgw** [comment]
@TheZoq2 There is some code in `src/driver.rs` that is commented out by default that helps trace some of the logic in this area. If you're comfortable compiling the program yourself, I can give a few more details. Alternatively, do you have a minimal example that demonstrates the issue?

**Comment by TheZoq2** [comment]
I'd be perfectly happy compiling and modifying the code. Right now, I can't seem to produce a minimal example without running into the original bug, but essentially, it seems to always rerun bibtex even if the bib file wasn't changed, which then triggers two more tex reruns

**Comment by pkgw** [comment]
@TheZoq2 If I'm understanding you correctly, the first half of what you're seeing is expected behavior. The reason is that the final document compilation requires a `bbl` file, which is what the `bibtex` tool creates from the `aux` and `bib` inputs. Tectonic doesn't save this file to disk by default, so even if no inputs change, it has to do the `tex / bibtex / tex / tex` dance to converge to the final document.

The overhead associated with the above is low enough that that it doesn't bother me, but if one wanted to minimize the number of passes, one would want to set up a [Ninja] file or Makefile to run Tectonic twice, the first time with arguments something like `--reruns=0 --keep-intermediates --hide=document.bbl`. (I haven't tested these arguments just now but something along those lines should be able to achieve the effect of avoiding rerunning bibtex.)

[Ninja]: https://ninja-build.org/

**Comment by TheZoq2** [comment]
> The overhead associated with the above is low enough that that it doesn't bother me
Doesn't rerunning bibtex force a rerun of latex which makes the compiliaton time `3*latex_time + bibtex` time instead of just `latex_time`

**Comment by pkgw** [comment]
@TheZoq2 Yes. It's certainly a large overhead in a relative sense, but in an absolute sense my document builds all generally take less than two seconds so I personally haven't been bothered. But I should be clear that other people are going to have other needs and we certainly want to support those! I do, however, think it's important that the "out-of-the-box" experience creates a PDF file from a LaTeX file with no reruns and no unnecessary support files left lying around on disk.

**Comment by TheZoq2** [comment]
> I do, however, think it's important that the "out-of-the-box" experience creates a PDF file from a LaTeX file with no reruns and no unnecessary support files left lying around on disk.

Agreed, I don't think this potential change should compromise that

> Yes. It's certainly a large overhead in a relative sense, but in an absolute sense my document builds all generally take less than two seconds so I personally haven't been bothered

In my case, I think it takes 3 seconds, which makes the total time slightly annoying

**Comment by doronbehar** [comment]
I have a similar / related issue to mention: Which is the fact that undefined citations warnings don't  appear in the default chatter. Warnings such as:

```
LaTeX Warning: Citation 'wikibook' on page 1 undefined on input line 5.
LaTeX Warning: Citation 'wombat2016' on page 1 undefined on input line 7.
LaTeX Warning: Citation 'lion2010' on page 1 undefined on input line 9.
```

[Discussed here](https://tectonic.newton.cx/t/what-is-your-recommended-workflow-for-tectonic-and-bibliography/270/2).

**Comment by pkgw** [comment]
Thanks for following up here. The proverbial "someone" should modify Tectonic's bibtex code to also emit its warnings through Tectonic's system so that they show up by default — it ought to be a relatively simple change to the C code.
